File: src-tauri/src/player.rs
use crate::{build_playlist, stop_current, AUDIO, CURRENT_INDEX, LAST_PATH, PLAYER, PLAYLIST};

use rodio::{Decoder, Sink};
use std::fs::File;
use std::io::BufReader;

#[tauri::command]
pub fn play_audio(path: String) -> Result<(), String> {
    log::debug!("Invoked play_audio function with path: {}", path);

    stop_current();

    let need_init = {
        let pl = PLAYLIST.lock().unwrap();

        pl.is_empty() || !pl.iter().any(|p| *p == path)
    };

    if need_init {
        build_playlist(&path)?;
    } else {
        let mut idx_guard = CURRENT_INDEX.lock().unwrap();
        let pl = PLAYLIST.lock().unwrap();

        if let Some(i) = pl.iter().position(|p| *p == path) {
            *idx_guard = i;
        }
    }

    let file = File::open(&path).map_err(|e| e.to_string())?;
    let source = Decoder::new(BufReader::new(file)).map_err(|e| e.to_string())?;
    let sink = Sink::try_new(&*AUDIO).map_err(|e| e.to_string())?;

    sink.append(source);
    sink.play();

    *PLAYER.lock().unwrap() = Some(sink);
    *LAST_PATH.lock().unwrap() = Some(path);

    Ok(log::info!("Audio playback started successfully"))
}

#[tauri::command]
pub fn pause_audio() -> Result<(), String> {
    log::debug!("Invoked pause_audio function");

    if let Some(sink) = &*PLAYER.lock().unwrap() {
        sink.pause();
    }

    Ok(log::info!("Audio playback paused successfully"))
}

#[tauri::command]
pub fn resume_audio() -> Result<(), String> {
    log::debug!("Invoked resume_audio function");

    if let Some(sink) = &*PLAYER.lock().unwrap() {
        sink.play();
    }

    Ok(log::info!("Audio playback resumed successfully"))
}

#[tauri::command]
pub fn restart_audio() -> Result<(), String> {
    log::debug!("Invoked restart_audio function");

    let path_opt = LAST_PATH.lock().unwrap().clone();

    if let Some(path) = path_opt {
        play_audio(path)
    } else {
        log::error!("No last track to restart");

        Err("No last track to restart".into())
    }
}

#[tauri::command]
pub fn next_audio() -> Result<(), String> {
    log::debug!("Invoked next_audio function");

    let path = {
        let playlist = PLAYLIST.lock().unwrap();

        if playlist.is_empty() {
            log::error!("Playlist is empty");
            return Err("Playlist is empty".into());
        }

        let mut index = CURRENT_INDEX.lock().unwrap();

        *index = (*index + 1) % playlist.len();

        playlist[*index].clone()
    };

    log::info!("Audio playback moved forward successfully");

    play_audio(path)
}

#[tauri::command]
pub fn previous_audio() -> Result<(), String> {
    log::debug!("Invoked previous_audio function");

    let path = {
        let playlist = PLAYLIST.lock().unwrap();

        if playlist.is_empty() {
            log::error!("Playlist is empty");
            return Err("Playlist is empty".into());
        }

        let mut index = CURRENT_INDEX.lock().unwrap();

        *index = if *index == 0 {
            playlist.len() - 1
        } else {
            *index - 1
        };

        playlist[*index].clone()
    };

    log::info!("Audio playback moved backward successfully");

    play_audio(path)
}

#[tauri::command]
pub fn get_current_audio() -> Option<String> {
    log::debug!("Invoked get_current_audio function");

    let playlist = PLAYLIST.lock().unwrap();
    let index = *CURRENT_INDEX.lock().unwrap();

    log::info!("Current audio got successfully");

    playlist.get(index).cloned()
}

File: src/lib/player.ts
import {cover, currentPath, isPlaying} from "@/stores/audio-store"
import {invoke} from "@tauri-apps/api/core"
import {debug, error, info} from "@tauri-apps/plugin-log"

export async function playAudio(path: string) {
    debug("Invoke play_audio function")

    try {
        await invoke("play_audio", {path})

        isPlaying.set(true)

        await loadCover(path)
    } catch (e) {
        error(`play_audio failed: ${e}`)
    }
}

export async function pauseAudio() {
    debug("Invoke pause_audio function")

    try {
        await invoke("pause_audio")

        isPlaying.set(false)
    } catch (e) {
        error(`pause_audio failed: ${e}`)
    }
}

export async function resumeAudio() {
    debug("Invoke resume_audio function")

    try {
        await invoke("resume_audio")

        isPlaying.set(true)
    } catch (e) {
        error(`resume_audio failed: ${e}`)
    }
}

export async function restartAudio() {
    debug("Invoke restart_audio function")

    try {
        await invoke("restart_audio")

        isPlaying.set(true)
    } catch (e) {
        error(`restart_audio failed: ${e}`)
    }
}

export async function nextAudio() {
    debug("Invoke next_audio function")

    try {
        await invoke("next_audio")
        await refreshCurrentAudio()

        isPlaying.set(true)
    } catch (e) {
        error(`next_audio failed: ${e}`)
    }
}

export async function previousAudio() {
    debug("Invoke previous_audio function")

    try {
        await invoke("previous_audio")
        await refreshCurrentAudio()

        isPlaying.set(true)
    } catch (e) {
        error(`previous_audio failed: ${e}`)
    }
}

export async function loadCover(path: string) {
    debug("Invoke get_audio_cover function")

    try {
        let invokedCover: string | null = await invoke("get_audio_cover", {path})

        if (invokedCover) {
            cover.set(invokedCover)

            info("Audio cover found")
        } else {
            cover.set(null)

            info("Audio cover not found")
        }
    } catch (e) {
        cover.set(null)

        error(`get_audio_cover failed: ${e}`)
    }
}

export async function refreshCurrentAudio() {
    debug("Invoke get_current_audio function")

    try {
        const path = await invoke<string | null>("get_current_audio")

        if (!path) {
            error("No current audio")
            return
        }

        currentPath.set(path)

        await loadCover(path)
    } catch (e) {
        error(`get_current_audio failed: ${e}`)
    }
}

export async function frontendReady() {
    debug("Invoke frontend_ready function")

    await invoke("frontend_ready")
}
